code-quality/code-validation.js

```diff
@@ -0,0 +1,492 @@
+        this.severity = "error";
+        this.applyToSections = ["awk"];
+
+        this.mark = function(content){
+            return content.replace(/writeDebug\(.+\)/gm, getSpan(this.severity, this.reason, "$&"));
+        }
+    },
+    "ifContainsSingleEqualSign": new function (){
+
+        // Single equal sign in an if is always true and the cause of bugs
+        // Example of an offending line: 
+        //if (variable = 1) { 
+
+        this.testName = "If statement with single equal sign";
+        this.reason = "Found an if statement contains a single equal sign. Since this is most likely an accident and it'd always return true it could cause strange bugs in the code. Consider replacing with double equal signs.";
+        this.severity = "error";
+        this.applyToSections = ["awk"];
+
+        this.mark = function(content){
+            return content.replace(/if\s*?\([^=]+[^=!]={1}[^=].+\)/gm, getSpan(this.severity, this.reason, "$&"));
+        }
+    },
+    "columnVariableManipulation": new function (){
+
+        // Changing column values is potentially the cause of bugs and should be avoided
+        // Example of an offending line:
+        // variable = test   
+
+        this.testName = "Column variable manipulation";
+        this.reason = "Changing column values (ie. $0, $1) could easily lead to unexpected behaviors.<br>It is highly recommended to instead save the column value to a variable and change that instead.";
+        this.severity = "warning";
+        this.applyToSections = ["awk"];
+
+        this.mark = function(content){
+            return content.replace(/g{0,1}sub.+?(\$[0-9])+/gm, getSpan(this.severity, this.reason, "$&"));
+        }
+    },
+    "trailingWhiteSpace": new function() {
+
+        //Trailing white space serves no purpose
+        // Example of an offending line:
+        // variable = test   
+
+        this.testName = "Trailing White-Space";
+        this.reason = "Trailing white space serves no purpose and should be removed";
+        this.severity = "error";
+        this.applyToSections = ["awk", "yaml"];
+
+        this.mark = function(content){
+            return content.replace(/[ \t]+$/gm, getSpan(this.severity, this.reason, "$&"))
+        }
+
+    },
+    "leadingTab": new function () {
+
+        // Tabs should not be used for indentation
+        // Example of an offending line:
+        //  variable = test
+
+        this.testName = "Leading tabs";
+        this.reason = "Tabs should not be used for indentation, please configure your editor to use space for indentation";
+        this.severity = "error";
+        this.applyToSections = ["awk", "yaml"];
+
+        this.mark = function(content){
+            return content.replace(/^[\t]+/gm, getSpan(this.severity, this.reason, "$&"))
+        }
+
+    },
+    "comparisonOperatorNoSpace": new function () {
```

Comment:
This is the only function I touched from the original file. Changed the name, tweaked the .reason, and refactored the pattern.